keps/976-plain-pods/README.md

```diff
@@ -0,0 +1,669 @@
+These Pods can be identified based on the ownerReference. For these pods:
+- The webhook should not add a scheduling gate.
+- The pod reconciler should not create a corresponding Workload object.
+
+Note that sometimes the Pods might not be directly owned by a known job object. Here are some
+special cases:
+- MPIJob: The launcher Pod is created through a batch/Job, which is also an known to Kueue, so
+  it's not an issue.
+- JobSet: Also creates Jobs, so not problematic.
+- RayJob: Pods are owned by a RayCluster object, which we don't currently support. This could be
+  hardcoded a known parent, or we could use label selectors for:
+  ```yaml
+  app.kubernetes.io/created-by: kuberay-operator
+  app.kubernetes.io/name: kuberay
+  ```
+
+#### Pods replaced on failure
+
+It is possible that users of plain Pods have a controller for them to handle failures and
+re-creations. These Pods should be able to use the quota that was already assigned to the Workload.
+
+Because Kueue can't know if Pods will be recreated or not, it will hold the entirety of the
+quota until it can determine that the whole Workload finished (all pods are terminated).
+In other words, Kueue won't support [dynamically reclaiming quota](https://github.com/kubernetes-sigs/kueue/issues/78)
+for plain Pods.
+
+### Risks and Mitigations
+
+#### Increased memory usage
+
+In order to support plain Pods, we need to start watching all Pods, even if they are not supposed
+to be managed by Kueue. This will increase the memory usage of Kueue just to maintain the
+informers.
+
+We can use the following mitigations:
+
+1. Drop the unused managedFields field from the Pod spec, like kube-scheduler is doing
+   https://github.com/kubernetes/kubernetes/pull/119556
+2. Filter out terminal Pods from informers, as they no longer influence quota usage
+   https://github.com/kubernetes/kubernetes/blob/99190634ab252604a4496882912ac328542d649d/pkg/scheduler/scheduler.go#L496
+
+## Design Details
+
+### Gating Pod Scheduling
+
+Pods subject to queueing should be prevented from scheduling until Kueue has admitted them in a
+specific flavor.
+
+Kubernetes 1.27 and newer provide the mechanism of [scheduling readiness](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)
+to prevent kube-scheduler from assigning Nodes to Pods.
+
+A Kueue webhook will inject to [Pods subject to queueing](#pods-subject-to-queueing):
+- A scheduling gate `kueue.x-k8s.io/admission` to prevent the Pod from scheduling.
+- A label `kueue.x-k8s.io/managed: true` so that users can easily identify pods that are/were
+  managed by Kueue.
+- A finalizer `kueue.x-k8s.io/managed` in order to reliably track pod terminations.
+
+A Pod reconciler will be responsible for removing the `kueue.x-k8s.io/admission` gate. If the Pods
+have other gates, they will remain Pending, but would be considered active from Kueue's perspective.
+
+#### Pods subject to queueing
```

Comment:
Should you maybe also consider pods owned by workloads to not be included in kueue?  

I worry that this configuration would be too limiting for a highly used cluster.  One that has deployments/statefulsets/daemon sets spread across many different namespaces.  I guess you could set up namespaces for all these ones but I could see that being difficult to keep up with.

Reply:
I added a PodSelector as well.
I don't want to plainly exclude Pods from (say) Deployments, as there could be valid reasons for queueing them.

Reply:
I am afraid I don't quite understand this.  I worry that if you have to explicitly remove items for Kueue it could be disruptive on a highly congested cluster.  

How will a PodSelector address this issue?

I'm envision a k8 cluster with a lot of users with a mix of services/batch.  A user has a namespace and they deploy services and batch jobs in that namespace.  Would it be simple for them to opt in on certain items for Kueue but avoid queueing services?

Reply:
Ah, I misunderstood your intent.

So far, we are not offering a mechanism for tracking quota usage of pods that were not queued (other than manually creating an admitted Workload). Let me add this to non-goals

I thought you were asking for a mechanism to exclude certain pods from queueing (and kueue overall) that was more granular than namespaces.

Reply:
Looks good!